File: Segundo_Curso/Segundo_Cuatri/PDOO/ruby/Irrgarten/Dice.rb
#encoding: UTF-8

module Irrgarten
	# Clase encargada de tomar las decisiones de azar en la partida
	class Dice
   	#Atributos de clase
		@@MAX_USES = 5			# Numero maximo de usos de armas y escudos
		@@MAX_INTELLIGENCE = 10.0	# Valor maximo para la inteligencia de jugadores y monstruos
		@@MAX_STRENGTH = 10.0		# Valor maximo para la fuerza de jugadores y monstruos
		@@RESURRECT_PROB = 0.3		# Numero maximo de armas recibidas al ganar un combate
		@@WEAPONS_REWARD = 2		# Probabilidad de que un jugador sea resucitado en cada turno
		@@SHIELDS_REWARD = 3		# Numero maximo de escudos recibidos al ganar un combate
		@@HEALTH_REWARD = 5		# Numero maximo de unidades de salud recibidas al ganar un combate
		@@MAX_ATTACK = 3		# Maxima potencia de armas
		@@MAX_SHIELD = 2		# Maxima potencia de escudos
		@@generator = Random.new	# Encargado de las decisiones de azar

		#Si a la funcion rand() se le pasa un float, devuelve un float entre 0 y ese, no inclusivo

		# No hay initialize creado por nosotros pues no se van a crear objetos

		# Metodo de clase encargado de generar una posicion en el tablero
		# @param max Posicion maxima de una dimension del tablero
		# @return Posicion en esa dimension del tablero
		def self.randomPos(max)
				@@generator.rand(max)
		end

		# Metodo de clase encargado de decidir que jugador empieza
		# @param players Numero de jugadores de la partida
		# @return Entero que representa el jugador que empieza
		def self.whoStarts(players)
				@@generator.rand(players)
		end

		# Metodo de clase encargado de determinar un valor de inteligencia
		# @return Valor de inteligencia
		def self.randomIntelligence
				(@@generator.rand(N))
		end

		# Metodo de clase encargado de determinar un valor de fuerza
		# @return Valor de fuerza
		def self.randomStrength
				(@@generator.rand(@@MAX_STRENGTH))
		end

		# Metodo de clase encargado de determinar una probabilidad de resurrecion
		# @return Valor de probabilidad de resurrecion
		def self.resurrectPlayer
				@@generator.rand(1.0) < @@RESURRECT_PROB
		end

		# Metodo de clase encargado de determinar un valor de la cantidad de armas que puede
		# recibir el jugador al final del combate
		# @return Valor de dicha cantidad
		def self.weaponsReward
				@@generator.rand(@@WEAPONS_REWARD)
		end

		# Metodo de clase encargado de determinar un valor de la cantidad de escudos que puede
		# recibir un jugador al final de un combate
		# @return Valor de dicha cantidad
		def self.shieldsReward
				@@generator.rand(@@SHIELDS_REWARD)
		end

		# Metodo de clase encargado de determinar la cantidad de unidades de salud que recibe un
		# jugador al terminar un combate
		# @return Valor de las unidades de salud
		def self.healthReward
				@@generator.rand(@@HEALTH_REWARD+1)
		end

		# Metodo de clase encargado de determinar un valor de poder de un arma
		# @return Poder del arma
		def self.weaponPower
				@@generator.rand(@@MAX_ATTACK.to_f+1)
		end

		# Metodo de clase encargado de determinar un valor de escudo
		# @return Valor de escudo
		def self.shieldPower
				@@generator.rand(@@MAX_SHIELD.to_f+1)
		end

		# Metodo de clase encargado de determinar el numero de usos que se le asignaran a un arma
		# o escudo
		# @return Usos de dichos objetos
		def self.usesLeft
				@@generator.rand(@@MAX_USES+1)
		end

		# Metodo de clase que devuelve la cantidad de competencia aplicada.
		# @return Competencia aplicada
		def self.intensity(competence)
				@@generator.rand(competence)
		end

		# Metodo de clase asigna la probabilidad de deshechar un objeto equipado
		# @return true si se descarta
		# 	  false si no se descarta
		def self.discardElement(usesLeft)
			(@@generator.rand() < (@@MAX_USES-usesLeft).to_f()/@@MAX_USES.to_f())
		end

		# Devuelve un movimiento segun la probabilidad en función de la inteligencia.
		# @param preference Movimiento preferente.
		# @param validMoves Posibles movimientos.
		# @param intelligence Valor de inteligencia que proporciona la probabilidad.
		# @return Movimiento de un fuzzy player.
		def self.nextStep(preference, validMoves,intelligence)
			if(intelligence >= @@generator.rand(@@MAX_INTELLIGENCE)) then 
				result = preference
			else
				if validMoves.nil? then
					result = nil
				else
					result = validMoves[@@generator.rand(validMoves.size)]
				end
			end
			result
		end
	end#class
end#module

File: Segundo_Curso/Segundo_Cuatri/PDOO/java/src/irrgarten/Dice.java
package irrgarten;

/*
 * Click nbfs://nbhost/SystemFileSystem/Templates/Licenses/license-default.txt to change this license
 * Click nbfs://nbhost/SystemFileSystem/Templates/Classes/Class.java to edit this template
 */

import java.util.Random;


/**
 *
 * @author airam
 */
public class Dice { 
    
    
    
    /**
     *  Numero máximo de usos de armas y escudos
     */
    static private int MAX_USES = 5; 
    /**
     *  Valor máximo para la inteligencia de jugadores y monstruos
     */
    static private float MAX_INTELLIGENCE = 10.0f; 
    /**
     *  Valor máximo fuerza jugadores
     */
    static private float MAX_STRENGTH =10.0f; 
    /**
     *  Probabilidad de ser resucitado en cada turno
     */
    static private float RESURRECT_PROB=0.3f; 
    /**
     *  Máximo de armas ganadas por combate ganado
     */
    static private int WEAPONS_REWARD=2; 
    /**
     *  Máximo de escudos ganados por combate ganado
     */
    static private int SHIELDS_REWARD=3;
    /**
     *  Maximo de vida recuperada por combate ganado
     */
    static private int HEALTH_REWARD=5;
    /**
     *  Máxima potencia de las armas
     */
    static private int MAX_ATTACK=3;
    /**
     *  Máximo potencia de los escudos
     */
    static private int MAX_SHIELD=2;
    
    /**
     *  Generador de números aleatorios
     */
    static Random generator;


    /**
     * Genera la posición relativa a la fila o columna inicial
     * @param max Cota superior del rango de valores a devolver, no inclusive.
     * @return El índice de la fila o columna
     */
    static public int randomPos(int max){
        return generator.nextInt(max);
    }
    
    /**
     * Decide qué jugador tendrá el primer turno
     * @param nPlayers Número de jugadores seleccionables
     * @return El índice del jugador elegido 
     */
    static public int whoStarts(int nPlayers){
        return generator.nextInt(nPlayers);
    }
    
    /**
     * Genera un valor aleatorio de inteligencia en el intervalo [0,MAX_INTELLIGENCE[
     * @return valor de inteligencia
     */
    static public float randomIntelligence(){
        return generator.nextFloat()*MAX_INTELLIGENCE;
    }
    
    /**
     * Genera un valor aleatorio de fuerza en el intervalo [0,MAX_STRENGTH[
     * @return valor de fuerza
     */
    static public float randomStrength(){
        return generator.nextFloat()*MAX_STRENGTH;
    }
    
    /**
     * Decide si un jugador fallecido debe resucitar en el turno actual
     * @return true si debe resucitar, false en caso contrario
     */
    static public boolean resurrectPlayer(){
        //Consideramos true como menor para ajustar bien los rangos
        return generator.nextFloat() < RESURRECT_PROB; 
    }
    
    /**
     * Decide la cantidad de armas ganadas tras un combate
     * @return el número de armas en el rango [0,WEAPONS_REWARD]
     */
    static public int weaponsReward(){
        //+1 para ajustar rango
        return generator.nextInt(WEAPONS_REWARD+1);
    }
    
    /**
     * Decide la cantidad de escudos ganadas tras un combate
     * @return el número de escudos en el rango [0,SHIELDS_REWARD]
     */
    static public int shieldsReward(){
        //+1 para ajustar rango
        return generator.nextInt(SHIELDS_REWARD+1);
    }
    
    /**
     * Decide la cantidad de vida recuperada tras un combate
     * @return La vida recuperada en el rango [0,HEALTH_REWARD]
     */
    static public int healthReward(){
        return generator.nextInt(HEALTH_REWARD+1);
    }
    
    //TODO: cuando sepas que hace expandir comentario
    /**
     * Decide el poder de un arma 
     * @return El poder del arma
     */
    static public float weaponPower(){
        return generator.nextFloat()*MAX_ATTACK;
    }

    /**
     * Decide la defensa de un escudo 
     * @return La defensa del arma
     */
    static public float shieldPower(){
        return generator.nextFloat()*MAX_SHIELD;
    }

    /**
     * Decide los usos restantes que tendrá un arma o escudo
     * @return Los usos restantes
     */
    static public int usesLeft(){
        return generator.nextInt(MAX_USES+1);
    }

    /**
     * Devuelve la cantidad de competencia aplicada con un valor aleatorio 
     * en el rango [0,competence[
     * @param competence La competencia
     * @return La competencia generada
     */
    static public float intensity(float competence){
        return generator.nextFloat() * competence;
    }

    /**
     * Decide si desechar un arma según el número de usos restantes
     * @param usesLeft Número de usos del objeto
     * @return True si decide tirar el arma, false en el caso contrario
     */
    static public boolean discardElement(int usesLeft){
        //distribuye de forma uniforme la probabilidad
        return generator.nextFloat() < (MAX_USES-usesLeft)/(float) MAX_USES;
    }
    
}
